README.md
# PROYECTO DE PROGRAMACION...
Proyecto de Programacion de Sistema de la ESPOL

## Instrucciones
El proyecto consiste en la implementacion de un servidor que envia y recibe archivos mediante la encriptacion de datos mediante uso de contrasena compartida utilizando el algoritmo de Diffie-Hellman, los comandos son los siguientes:
GET <nombreArchivo> garantiza la descarga del archivo al servidor
PUT <nombreArchivo> <longitudArchivo> sube el archivo al servidor enviando conjuntamente el TAMANO DEL ARCHIVO.- Para nuestro ejemplo tambien se le tiene que pasar el tamano del archivo, esto es un requisito importante sin embargo se esta conciente que podria estar a sueltas de mejoras.
READY confirma que esta listo para la descarga del archivo
CHAO ejecuta la descomunicacion entre pares
LIST enlista los archivos existente en el proyecto
Este proyecto se baso bajo la modalidad de multi-hilos y mantiene una libreria de forma estatica.

Toda comunicación entre cliente y servidor debe ser enviada en modo ASCII con salto de línea al final, excepto la transmisión del contenido del archivo, el cual debe ser enviado mediante blowfish pero guardado en el disco duro en ASCII.

El programa servidor *server* tiene el siguiente comportamiento:
```
$ ./server -h
Servidor simple de descarga de archivos.
uso:
 ./server <puerto>
 ./server -h
 ./server -d <puerto>
Opciones:
 -h			Ayuda, muestra este mensaje
 -d 			Activa el modo Daemon
```

El programa cliente *client* tiene el siguiente comportamiento:
```
$ ./client -h
Cliente simple de descarga de archivos.
uso:
 ./client <hostname> <puerto>
 ./client -h
Opciones:
 -h			Ayuda, muestra este mensaje
```



## Compilación

```

Para compilar cliente y servidor:
```
NAME=crypto
DEPS = common.h
CFLAGS = -Wall -c
SLIB = lib$(NAME).a
LDIR= libs
LIBS=libs/archivos/
SSTC= -Wl, -Bstatic

.PHONY: all debug sanitize clean

all: $(LDIR)/$(SLIB) server client

$(LDIR)/$(SLIB): $(LIBS)blowfish.o $(LIBS)uECC.o $(LIBS)sha256.o $(LIBS)common.o
	ar rcs $(LDIR)/$(SLIB) $(LIBS)blowfish.o $(LIBS)uECC.o $(LIBS)sha256.o $(LIBS)common.o

$(LIBS)blowfish.o: blowfish.c
	gcc $(CFLAGS) -c $(LIBS)blowfish.c -o $(LIBS)blowfish.o

$(LIBS)uECC.o: uECC.c
	gcc $(CFLAGS) -c $(LIBS)uECC.c -o $(LIBS)uECC.o

$(LIBS)sha256.o: sha256.c
	gcc $(CFLAGS) -c $(LIBS)sha256.c -o $(LIBS)sha256.o 
 
server: server.c common.o  $(DEPS) 
	gcc  -g -o server server.c  common.o $(SSTC) $(LDIR)/$(SLIB)  -lpthread

client: client.c common.o $(DEPS)
	gcc  -g -o client client.c  common.o $(SSTC) $(LDIR)/$(SLIB)  #-lpthread

common.o: common.c $(DEPS)
	gcc $(CFLAGS) -c common.c

```

# Compila usando la opción -g para facilitar la depuración con gdb.
debug: DFLAGS = -g
debug: clean all

# Compila habilitando la herramienta AddressSanitizer para
# facilitar la depuración en tiempo de ejecución.
sanitize: DFLAGS = -fsanitize=address,undefined
sanitize: clean all

clean:
	rm -rf server client only_server *.o lib/archivos/*.o 
```
Para compilar cliente y servidor facilitando la depuración con gdb:
```
# Compila usando la opción -g para facilitar la depuración con gdb.
.PHONY: all debug sanitize clean
debug: DFLAGS = -g
debug: clean all

$ make debug
```
Para compilar cliente y servidor habilitando la herramienta AddressSanitizer, facilita la depuración en tiempo de ejecución:
```
sanitize: DFLAGS = -fsanitize=address,undefined
sanitize: clean all
$ make sanitize
```
## Integrantes

```
Morales Mendoza Jose
  Actividades realizadas:
	-Agregar Permisos de escritura y lectura al archivo que recibe el cliente.
	-Encriptacion de archivos
	-Claves mediante modelo Diffie-Helman
	-Metodo GET
	-Edicion de Makefile
	-Modo Daemon
```

```
Vargas Aguirre Lilibeth
  Actividades realizadas:
	-Edición del README.
	-Manejo de directorios de archivos clientes y servidor.
	-Metodo PUT, CHAO, READY
	-Automatas
	-Validaciones e implementacion de Hilos
	-Edicion de Makefile
	
```
# Amazon Web Service 
	https://user-images.githubusercontent.com/51379487/130611144-1c1f92a8-57fe-485d-92d5-ee0983655e63.png
```
Para el nuevo diseño se adicionó servicios de AWS como:
	 Amazon S3 bucket: utilizado especialmente para el almacenamiento de los archivos encriptados que ya sea fueron descargados por el usuario o subidos.
	 IAM: administra el acceso de AWS de forma segura, nos permite crear usuarios y permisos para acceder a los recursos.
	Role de IAM: permite tener políticas y permisos específicos para las entidades, asigna credenciales temporales para la sesión del rol. [3]
	 Temporary security credential: nos brinda acceso al bucket s3 o cualquier servicio de Amazon que se sincronice con el rol.
	 ECR: servidor administrado de registro de Docker de AWS nos permite insertar, extraer imágenes de repositorios privados de Docker con permisos de
	nivel de recursos compatible con IAM, de esta manera las instancias de EC2 tengan acceso a los repositorios.
	 Taks de ECR: permite ejecutar la imagen de Docker
	 ECS cluster: un clúster con finalidad de que agrupar instancias de contenedores, permite escalar aplicaciones con contenedores, se administra el número de 			instancias EC2.
	 Imagen de Docker: una imagen de Docker permite contener el código de la aplicación.
	 EC2: Amazon Elastic Compute Cloud es una instancia de GPU que ejecuta una tarea ECS, provee facultad informática de forma segura y de tamaño adaptable.
```


fileClient/CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

# [1.351.0] - 2021-08-19
### Added
- Added new warehouse administrator.


# [1.350.6] - 2021-08-17
### Fixed
- Fixed error when updating a withholding if the support document date is incorrect.
- Fixed error when getting the information of a client in accounts receivable report.


# [1.350.5] - 2021-08-12
### Fixed
- Fixed account owner id type generated in batch file for recurring debit.


# [1.350.4] - 2021-08-06
### Fixed
- Fixed error when recording a payment with Stripe if the invoice has more than one customer email.
- Fixed error when disabling a payment method.


# [1.350.3] - 2021-08-05
### Changed
- Enabled to save account owner identification and name when registering for recurring debit.


# [1.350.2] - 2021-08-04
### Fixed
- Fixed error in accounts receivable report when getting the information of a client.


# [1.350.1] - 2021-08-02
### Fixed
- Fixed error when disabling a payment method that caused it to be deleted in the two configured environments.
- Fixed error when configuring a payment method.


# [1.350.0] - 2021-07-30
### Added
- Added info modal about referrals program use.


# [1.349.3] - 2021-07-27
### Fixed
- Fixed error when saving an invoice with withholdings values.
- Fixed error when recording a payphone payment, it now supports correctly idempotency.


# [1.349.2] - 2021-07-23
### Fixed
- Fixed stripe payment properties.


# [1.349.1] - 2021-07-21
### Added
- Added publishable key to stripe credentials.


# [1.349.0] - 2021-07-21
### Added
- Added stripe to the list of payment methods for invoice.


# [1.348.0] - 2021-07-20
### Added
- Added help information about referral state in referral section.
- Deleted referral prize state column in referral section.
- Added support in datatable component to add help tooltips in headers.


# [1.347.0] - 2021-07-19
### Added
- Improves the option to import received documents.


# [1.346.1] - 2021-07-15
### Fixed
- Fixed overrided reponse variable on register referral invocation.


# [1.346.0] - 2021-07-12
### Added
- Added PayPhone to the list of payment methods for invoices.


# [1.345.4] - 2021-07-07
### Fixed
- Fixed error when creating a quote, the sequence is not as expected.

### Added
- Added sku product when searching and adding an item in quotes section


# [1.345.3] - 2021-07-07
### Fixed
- Fixed error in debit order section. Pagination is working correctly now.


# [1.345.2] - 2021-07-06
### Fixed
- Fixed error when issuing an invoice or credit note, if status is not autorized inventory transaction is generated anyway.


# [1.345.1] - 2021-07-05
### Fixed
- Fixed wrong argument name in register referral function.

# [1.345.0] - 2021-07-02
### Added
- New Referrals program section on datil panel.
- Form to invite contacts on referrals program section.
- Referrals list on referrals program section.
- DatilOS services to invite, list, register referrals.

# [1.344.4] - 2021-06-30
### Fixed
- Fixed error in batch issue section for connect accounts, the option to download the report was not displaying information.
- Fixed error in credit note batch issue, the option to see credit notes belonging to a batch was not displaying information.
- Fixed error when issuing documents.


## [1.344.3] - 2021-06-24
### Fixed
- Fixed error when canceling an adjustment transaction.
- Fixed error when filtering purchases settlements and credit notes by batch.


## [1.344.2] - 2021-06-24
### Fixed
- Fixed credit validations allowing invoices being created when the credit exceeds the invoice's balance.


## [1.344.1] - 2021-06-22
### Fixed
- Fixed download feature when the custom template feature flag is enabled


## [1.344.0] - 2021-06-21
### Added
- Added option to issue batch invoices for connect accounts.


## [1.343.0] - 2021-06-18
### Added
- Process to import Banco Pichincha debit transactions.


## [1.342.0] - 2021-06-16
### Added
- Added field to specify payment method in invoices with credit


## [1.341.2] - 2021-06-16
### Fixed
- Makes item "codigo_auxiliar" field in purchase settlements not required.


## [1.341.1] - 2021-06-15
### Fixed
- Makes supplier address field in purchase settlements not required as it should be.


## [1.341.0] - 2021-06-15
### Added
- Added option to issue batch purchases settlements.


## [1.340.5] - 2021-06-15
### Fixed
- Displays error message when issuing a document in batch.
- Fixes error when filtering by sales representative in sales report.


## [1.340.4] - 2021-06-14
### Fixed
- Improves perfomance in the sales report


## [1.340.3] - 2021-06-12
### Fixed
- Added restrictions when accessing the sales report
- Fixed error when exporting the sales report.


## [1.340.2] - 2021-06-11
### Fixed
- Fixed error when filtering in the sales report.


## [1.340.1] - 2021-06-11
### Fixed
- Fixed sales report, discounts were not being considered and it always was filtering by test environment.


## [1.340.0] - 2021-06-10
### Added
- Added customer, product, sales representative and category filter in the sales report.

### Fixed
- Fixed error in sales report.


## [1.339.2] - 2021-06-10
### Added
- Added category when downloading the sales by items report.


## [1.339.1] - 2021-06-09
### Fixed
- Fixes bug in generated debit account file.


## [1.339.0] - 2021-06-08
### Added
- New option to generate bank account debit orders.


## [1.338.1] - 2021-06-07
### Fixed
- Fixed error when issuing a credit note with items with and without warehouse, inventory transaction was not generated.


## [1.337.14] - 2021-06-05
### Fixed
- Fixed error when issuing a credit note, inventory transaction was not generated.


## [1.337.13] - 2021-06-03
### Fixed
- Fixed error when issuing an invoice with sales representative.


## [1.337.12] - 2021-06-02
### Added
- Added salesrep when issuing an invoice via API.

### Fixed

libs/archivos/sha256.h
/*********************************************************************
* Filename:   sha256.h
* Author:     Brad Conte (brad AT bradconte.com)
* Copyright:
* Disclaimer: This code is presented "as is" without any guarantees.
* Details:    Defines the API for the corresponding SHA1 implementation.
*********************************************************************/

#ifndef SHA256_H
#define SHA256_H

/*************************** HEADER FILES ***************************/
#include <stddef.h>
#include <getopt.h>

/****************************** MACROS ******************************/
#define SHA256_BLOCK_SIZE 32            // SHA256 outputs a 32 byte digest

/**************************** DATA TYPES ****************************/
typedef unsigned char BYTE;             // 8-bit byte
typedef unsigned int  WORD;             // 32-bit word, change to "long" for 16-bit machines

typedef struct {
	BYTE data[64];
	WORD datalen;
	unsigned long long bitlen;
	WORD state[8];
}SHA256_CTX;

/*********************** FUNCTION DECLARATIONS **********************/
void sha256_init(SHA256_CTX *ctx);
void sha256_update(SHA256_CTX *ctx, const BYTE data[], size_t len);
void sha256_final(SHA256_CTX *ctx, BYTE hash[]);

#endif   // SHA256_H


libs/archivos/blowfish.h
/*********************************************************************
* Filename:   blowfish.h
* Author:     Brad Conte (brad AT bradconte.com)
* Copyright:
* Disclaimer: This code is presented "as is" without any guarantees.
* Details:    Defines the API for the corresponding Blowfish implementation.
*********************************************************************/

#ifndef BLOWFISH_H
#define BLOWFISH_H

/*************************** HEADER FILES ***************************/
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdbool.h>
#include <string.h>
#include <getopt.h>

/****************************** MACROS ******************************/
#define BLOWFISH_BLOCK_SIZE 8           // Blowfish operates on 8 bytes at a time

/**************************** DATA TYPES ****************************/
typedef unsigned char BYTE;             // 8-bit byte
typedef unsigned int  WORD;             // 32-bit word, change to "long" for 16-bit machines

typedef struct {
   WORD p[18];
   WORD s[4][256];
} BLOWFISH_KEY;

/*********************** FUNCTION DECLARATIONS **********************/
void blowfish_key_setup(const BYTE user_key[], BLOWFISH_KEY *keystruct, size_t len);
void blowfish_encrypt(const BYTE in[], BYTE out[], const BLOWFISH_KEY *keystruct);
void blowfish_decrypt(const BYTE in[], BYTE out[], const BLOWFISH_KEY *keystruct);
BYTE* conv32bytes(char *clave);
char * copiarNombre(char *nombreFile);

#endif   // BLOWFISH_H


libs/archivos/sha256.c
/*********************************************************************
* Filename:   sha256.c
* Author:     Brad Conte (brad AT bradconte.com)
* Copyright:
* Disclaimer: This code is presented "as is" without any guarantees.
* Details:    Implementation of the SHA-256 hashing algorithm.
              SHA-256 is one of the three algorithms in the SHA2
              specification. The others, SHA-384 and SHA-512, are not
              offered in this implementation.
              Algorithm specification can be found here:
               * http://csrc.nist.gov/publications/fips/fips180-2/fips180-2withchangenotice.pdf
              This implementation uses little endian byte order.
*********************************************************************/

/*************************** HEADER FILES ***************************/
#include <stdlib.h>
#include <memory.h>
#include <stdio.h>
#include <string.h>
#include "sha256.h"

/****************************** MACROS ******************************/
#define ROTLEFT(a,b) (((a) << (b)) | ((a) >> (32-(b))))
#define ROTRIGHT(a,b) (((a) >> (b)) | ((a) << (32-(b))))

#define CH(x,y,z) (((x) & (y)) ^ (~(x) & (z)))
#define MAJ(x,y,z) (((x) & (y)) ^ ((x) & (z)) ^ ((y) & (z)))
#define EP0(x) (ROTRIGHT(x,2) ^ ROTRIGHT(x,13) ^ ROTRIGHT(x,22))
#define EP1(x) (ROTRIGHT(x,6) ^ ROTRIGHT(x,11) ^ ROTRIGHT(x,25))
#define SIG0(x) (ROTRIGHT(x,7) ^ ROTRIGHT(x,18) ^ ((x) >> 3))
#define SIG1(x) (ROTRIGHT(x,17) ^ ROTRIGHT(x,19) ^ ((x) >> 10))

/**************************** VARIABLES *****************************/
static const WORD k[64] = {
	0x428a2f98,0x71374491,0xb5c0fbcf,0xe9b5dba5,0x3956c25b,0x59f111f1,0x923f82a4,0xab1c5ed5,
	0xd807aa98,0x12835b01,0x243185be,0x550c7dc3,0x72be5d74,0x80deb1fe,0x9bdc06a7,0xc19bf174,
	0xe49b69c1,0xefbe4786,0x0fc19dc6,0x240ca1cc,0x2de92c6f,0x4a7484aa,0x5cb0a9dc,0x76f988da,
	0x983e5152,0xa831c66d,0xb00327c8,0xbf597fc7,0xc6e00bf3,0xd5a79147,0x06ca6351,0x14292967,
	0x27b70a85,0x2e1b2138,0x4d2c6dfc,0x53380d13,0x650a7354,0x766a0abb,0x81c2c92e,0x92722c85,
	0xa2bfe8a1,0xa81a664b,0xc24b8b70,0xc76c51a3,0xd192e819,0xd6990624,0xf40e3585,0x106aa070,
	0x19a4c116,0x1e376c08,0x2748774c,0x34b0bcb5,0x391c0cb3,0x4ed8aa4a,0x5b9cca4f,0x682e6ff3,
	0x748f82ee,0x78a5636f,0x84c87814,0x8cc70208,0x90befffa,0xa4506ceb,0xbef9a3f7,0xc67178f2
};

/*********************** FUNCTION DEFINITIONS ***********************/
void sha256_transform(SHA256_CTX *ctx, const BYTE data[])
{
	WORD a, b, c, d, e, f, g, h, i, j, t1, t2, m[64];

	for (i = 0, j = 0; i < 16; ++i, j += 4)
		m[i] = (data[j] << 24) | (data[j + 1] << 16) | (data[j + 2] << 8) | (data[j + 3]);
	for ( ; i < 64; ++i)
		m[i] = SIG1(m[i - 2]) + m[i - 7] + SIG0(m[i - 15]) + m[i - 16];

	a = ctx->state[0];
	b = ctx->state[1];
	c = ctx->state[2];
	d = ctx->state[3];
	e = ctx->state[4];
	f = ctx->state[5];
	g = ctx->state[6];
	h = ctx->state[7];

	for (i = 0; i < 64; ++i) {
		t1 = h + EP1(e) + CH(e,f,g) + k[i] + m[i];
		t2 = EP0(a) + MAJ(a,b,c);
		h = g;
		g = f;
		f = e;
		e = d + t1;
		d = c;
		c = b;
		b = a;
		a = t1 + t2;
	}

	ctx->state[0] += a;
	ctx->state[1] += b;
	ctx->state[2] += c;
	ctx->state[3] += d;
	ctx->state[4] += e;
	ctx->state[5] += f;
	ctx->state[6] += g;
	ctx->state[7] += h;
}

void sha256_init(SHA256_CTX *ctx)
{
	ctx->datalen = 0;
	ctx->bitlen = 0;
	ctx->state[0] = 0x6a09e667;
	ctx->state[1] = 0xbb67ae85;
	ctx->state[2] = 0x3c6ef372;
	ctx->state[3] = 0xa54ff53a;
	ctx->state[4] = 0x510e527f;
	ctx->state[5] = 0x9b05688c;
	ctx->state[6] = 0x1f83d9ab;
	ctx->state[7] = 0x5be0cd19;
}

void sha256_update(SHA256_CTX *ctx, const BYTE data[], size_t len)
{
	WORD i;

	for (i = 0; i < len; ++i) {
		ctx->data[ctx->datalen] = data[i];
		ctx->datalen++;
		if (ctx->datalen == 64) {
			sha256_transform(ctx, ctx->data);
			ctx->bitlen += 512;
			ctx->datalen = 0;
		}
	}
}

void sha256_final(SHA256_CTX *ctx, BYTE hash[])
{
	WORD i;

	i = ctx->datalen;

	// Pad whatever data is left in the buffer.
	if (ctx->datalen < 56) {
		ctx->data[i++] = 0x80;
		while (i < 56)
			ctx->data[i++] = 0x00;
	}
	else {
		ctx->data[i++] = 0x80;
		while (i < 64)
			ctx->data[i++] = 0x00;
		sha256_transform(ctx, ctx->data);
		memset(ctx->data, 0, 56);
	}

	// Append to the padding the total message's length in bits and transform.
	ctx->bitlen += ctx->datalen * 8;
	ctx->data[63] = ctx->bitlen;
	ctx->data[62] = ctx->bitlen >> 8;
	ctx->data[61] = ctx->bitlen >> 16;
	ctx->data[60] = ctx->bitlen >> 24;
	ctx->data[59] = ctx->bitlen >> 32;
	ctx->data[58] = ctx->bitlen >> 40;
	ctx->data[57] = ctx->bitlen >> 48;
	ctx->data[56] = ctx->bitlen >> 56;
	sha256_transform(ctx, ctx->data);

	// Since this implementation uses little endian byte ordering and SHA uses big endian,
	// reverse all the bytes when copying the final state to the output hash.
	for (i = 0; i < 4; ++i) {
		hash[i]      = (ctx->state[0] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 4]  = (ctx->state[1] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 8]  = (ctx->state[2] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 12] = (ctx->state[3] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 16] = (ctx->state[4] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 20] = (ctx->state[5] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 24] = (ctx->state[6] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 28] = (ctx->state[7] >> (24 - i * 8)) & 0x000000ff;
	}
}





